mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala

```diff
@@ -275,16 +276,13 @@ object MatrixFactorizationModel extends Loader[MatrixFactorizationModel] {
       num: Int): RDD[(Int, Array[(Int, Double)])] = {
     val srcBlocks = blockify(rank, srcFeatures)
     val dstBlocks = blockify(rank, dstFeatures)
+    val output = new ArrayBuffer[(Int, (Int, Double))]()
     val ratings = srcBlocks.cartesian(dstBlocks).flatMap {
       case ((srcIds, srcFactors), (dstIds, dstFactors)) =>
-        val m = srcIds.length
-        val n = dstIds.length
         val ratings = srcFactors.transpose.multiply(dstFactors)
-        val output = new Array[(Int, (Int, Double))](m * n)
-        var k = 0
+        output.clear()
         ratings.foreachActive { (i, j, r) =>
```

Comment:
We don't need `output` to hold the buffer. The following should work, though it doesn't really fix the GC problem:

~~~scala
        for (i <- 0 until m; j <- 0 until n) yield {
          (srcIds(i), dstIds(j), ratings(i, j))
        }
~~~